server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -881,24 +881,30 @@ private void findAliasInfo(final String[] aliases, final String[] possibleMatche
 
         boolean matchAllAliases = patterns.length == 0;
 
+        // memoize pattern match against aliases to avoid repeatedly matching when multiple indices share an alias
+        HashMap<String, Boolean> seenAliases = new HashMap<>();
```

Comment:
Are we at all worried that this HashMap could get very large ?

Reply:
It's not impossible, but the caller of this function is already building and [returning a map](https://github.com/elastic/elasticsearch/blob/57846444cc185f3a9a34c068ed566eb462eb565f/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L811) that is roughly the same size (number of keys up to the sum of aliases across all concrete indices, with keys being pointers to existing strings and values being new boxes). So ~2x transient usage? The entries themselves are pretty small (pointer, boxed bool) so you'd need a lot to matter in absolute terms I think.

Having said that I don't know whether we have any limits or typical upper bounds on the number of aliases in a cluster. I also don't know whether we have any limits on the number of concurrent GetAliases calls that we admit. Either of those could I suppose make the memory usage of this map matter, though my guess is that we'd be in a somewhat pathological environment at this point, where we were having other problems not related to the size of this allocation.

I suppose if we needed to we could bound this thing as some sort of LRU but I would guess that we'd have to then know how to size it, which depends probably on answers to the questions in the paragraph above about total number of aliases and max concurrency.